compiler_opt/rl/feature_ops.py

```diff
@@ -80,3 +85,88 @@ def normalize(obs: types.Float):
     return tf.concat(features, axis=-1)
 
   return normalize
+
+
+def get_ir2vec_normalize_fn(ir2vec_with_z_score_normalization: bool = False,
+                            eps: float = 1e-8):
+  """Return a normalization function for embeddings.
+
+  Args:
+    ir2vec_with_z_score_normalization: If True, applies z-score standardization
+      (whitening) to the embeddings per batch. This transforms the embeddings
+      to have zero mean and unit variance across the batch dimension.
+      Setting this to True would reduce the impact of outlier embeddings and
+      improve training stability, especially when embeddings from different
+      batches have significantly different scales.
+
+    eps: Small epsilon value added to standard deviation to prevent division
+      by zero during standardization.
+
+  Returns:
+    A function that normalizes IR2Vec embeddings according to the specified
+    parameters.
+  """
+  if ir2vec_with_z_score_normalization:
+    # Whitens the embeddings per batch.
+    def z_score_normalize(batch_embeddings: types.Float):
+      mean = tf.math.reduce_mean(batch_embeddings, axis=0, keepdims=True)
+      std = tf.math.reduce_std(batch_embeddings, axis=0, keepdims=True)
+      standardized_embeddings = (batch_embeddings - mean) / (std + eps)
+      return standardized_embeddings
+
+    return z_score_normalize
+  # Currently, we just return the identity function for embeddings.
+  # We can extend this to include other normalizations like L2
+  # normalization if needed.
+  return lambda x: identity_fn(x, expand_dims=False)
+
+
+def get_ir2vec_dimensions_from_vocab_file(vocab_file_path: str) -> int:
+  """Read the IR2Vec vocabulary file and get embedding dimensions from the
+  first embedding in the first section.
+
+  Args:
+    vocab_file_path: Path to the IR2Vec vocabulary JSON file.
+
+  Returns:
+    The number of dimensions in the embeddings.
+
+  Raises:
+    tf.errors.NotFoundError: If the vocabulary file cannot be found.
+    json.JSONDecodeError: If the vocabulary file contains invalid JSON.
```

Comment:
I'm not sure you should list these two here. There's sort of an implict contract that the user is providing a valid JSON file.

>  You should not document exceptions that get raised if the API specified in the docstring is violated (because this would paradoxically make behavior under violation of the API part of the API).

https://google.github.io/styleguide/pyguide.html#doc-function-raises

Relevant section of the style guide.